flink-yarn/src/test/java/org/apache/flink/yarn/YarnClusterClientProviderTest.java

```diff
@@ -0,0 +1,130 @@
+ * limitations under the License.
+ */
+
+package org.apache.flink.yarn;
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.JobManagerOptions;
+import org.apache.flink.configuration.RestOptions;
+import org.apache.flink.yarn.configuration.YarnConfigOptions;
+
+import org.apache.commons.beanutils.ConvertUtils;
+import org.apache.hadoop.yarn.api.records.ApplicationAttemptId;
+import org.apache.hadoop.yarn.api.records.ApplicationId;
+import org.apache.hadoop.yarn.api.records.ApplicationReport;
+import org.apache.hadoop.yarn.api.records.YarnApplicationState;
+import org.apache.hadoop.yarn.util.Records;
+import org.junit.Test;
+
+import static org.apache.hadoop.yarn.api.records.FinalApplicationStatus.UNDEFINED;
+import static org.junit.Assert.assertEquals;
+
+/** Test for {@link YarnClusterClientProvider}. */
+public class YarnClusterClientProviderTest {
+
+    /** Once Yarn app master has been running, it should not retrieve app report again. */
+    @Test
+    public void testNoLongerRetrieveAppReportOnceAppRunning() {
+        final Configuration flinkConf = new Configuration();
+        final ApplicationId applicationId =
+                ApplicationId.newInstance(System.currentTimeMillis(), 1);
+
+        final ApplicationReportProvider mockAppReportProvider =
+                MockApplicationReportProvider.of(applicationId, "localhost", 10);
+        assertEquals(0, ((MockApplicationReportProvider) mockAppReportProvider).getInvokeNumber());
+
+        YarnClusterClientProvider provider =
+                YarnClusterClientProvider.of(mockAppReportProvider, flinkConf, applicationId);
+
+        provider.getClusterClient();
+        assertEquals(1, ((MockApplicationReportProvider) mockAppReportProvider).getInvokeNumber());
+
+        provider.getClusterClient();
+        assertEquals(1, ((MockApplicationReportProvider) mockAppReportProvider).getInvokeNumber());
+    }
+
+    @Test
+    public void testJMHostShouldSetIntoConfig() {
+        final Configuration flinkConf = new Configuration();
+        final ApplicationId applicationId =
+                ApplicationId.newInstance(System.currentTimeMillis(), 1);
+        final String host = "localhost";
+        final int rpcPort = 10000;
+        final ApplicationReportProvider mockAppReportProvider =
+                MockApplicationReportProvider.of(applicationId, host, rpcPort);
+
+        YarnClusterClientProvider.of(mockAppReportProvider, flinkConf, applicationId)
+                .getClusterClient();
+
+        assertEquals(flinkConf.get(JobManagerOptions.ADDRESS), host);
+        assertEquals(flinkConf.getInteger(JobManagerOptions.PORT), rpcPort);
+        assertEquals(flinkConf.get(RestOptions.ADDRESS), host);
+        assertEquals(flinkConf.getInteger(RestOptions.PORT), rpcPort);
+        assertEquals(
+                flinkConf.get(YarnConfigOptions.APPLICATION_ID),
+                ConvertUtils.convert(applicationId));
```

Comment:
I think this test relies on the fact that `YarnClusterClientProvider` modifies the given `flinkConf`. This is an implementation detail. I think it is better to use the public API for assertions. If this is not possible then we could think about adding a package private `getRestClusterClientConfiguration()` or so that can be retrieved from the provider. But also this method leaks implementation details.

Reply:
No need to add extra component. Just get conf from clusterClient like
```
YarnClusterClientProvider provider =
                YarnClusterClientProvider.of(mockAppReportProvider, flinkConf, applicationId);

        ClusterClient clusterClient = provider.getClusterClient();

        final Configuration clusterClientConf = clusterClient.getFlinkConfiguration();
```